schemachanger: remove RBR fallback for add column

annrpom opened the pull request:
This patch removes the regional by row fallback for `ADD COLUMN`
in the declarative schema changer.

Epic: [CRDB-31282](https://cockroachlabs.atlassian.net/browse/CRDB-31282)
Fixes: https://github.com/cockroachdb/cockroach/issues/80545

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140617)
<!-- Reviewable:end -->

annrpom committed:
schemachanger: remove RBR fallback for add column

This patch removes the regional by row fallback for `ADD COLUMN`
in the declarative schema changer.

Epic: CRDB-31282
Fixes: #80545

Release note: None

rafiss reviewed (changes requested):
this is super exciting!! great work.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140617#-:-OLApL541wTbHl7-ewjO:b-c3pixd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_regional_by_row/add_column_regional_by_row.definition` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140617#-OLApCX1BwCHULp_ZHIV:-OLApCX1BwCHULp_ZHIW:b-yrwf3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d13d41f499fdb7cf8c87a9677c805486de8becf9/pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_regional_by_row/add_column_regional_by_row.definition#L2)):*
> ```
> setup
> CREATE DATABASE multiregion_db PRIMARY REGION "us-east1" REGIONS "us-east2", "us-east3" SURVIVE REGION FAILURE;
> ```

should we also have a test with SURVIVE ZONE FAILURE, or do you feel like this is sufficient? 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140617#-OLAoOApBjNsMO76bYDP:-OLAoOApBjNsMO76bYDQ:b-dn5wcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d13d41f499fdb7cf8c87a9677c805486de8becf9/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L128)):*
> ```Go
> 	if idx != nil {
> 		if isRBR {
> 			panicIfRegionChangeUnderwayOnRBRTable(b, "add a UNIQUE COLUMN", tbl.TableID)
> ```

nit: let's remove the `if isRBR` check here. panicIfRegionChangeUnderwayOnRBRTable already has that check in it.

also is the `idx != nil` check too broad? previously, we'd only call panicIfRegionChangeUnderwayOnRBRTable if a UNIQUE column was being added.


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140617)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140617#-OLAoOApBjNsMO76bYDP:-OLBH4rI4Q9qBiB69609:b3ymlyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d13d41f499fdb7cf8c87a9677c805486de8becf9/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L128)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: let's remove the `if isRBR` check here. panicIfRegionChangeUnderwayOnRBRTable already has that check in it.

also is the `idx != nil` check too broad? previously, we'd only call panicIfRegionChangeUnderwayOnRBRTable if a UNIQUE column was being added.
</blockquote></details>

re:nit -- ah oki right

re: broadness of `idx != nil` check -- i aligned this with legacy's implementation https://github.com/cockroachdb/cockroach/blob/42b9060871865023e97bbbe2e42ada6a8cec55d3/pkg/sql/add_column.go#L79-L92; iiuc this should still be equivalent since primary keys are also unique, but lmk if that's not what it going on here https://github.com/cockroachdb/cockroach/blob/4ab689e1d9f5420c6cef41de6808455ed55e2a14/pkg/sql/catalog/tabledesc/table.go#L42-L44
___
*[`pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_regional_by_row/add_column_regional_by_row.definition` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140617#-OLApCX1BwCHULp_ZHIV:-OLBAK1b7d8EpOlDALQW:b-ux71jj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d13d41f499fdb7cf8c87a9677c805486de8becf9/pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_regional_by_row/add_column_regional_by_row.definition#L2)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should we also have a test with SURVIVE ZONE FAILURE, or do you feel like this is sufficient?
</blockquote></details>

i feel like this is sufficient -- at least for a DSC end_to_end test, but what were you thinking about testing? the zone config generated from a `SURVIVE ZONE FAILURE` (which is code i touched, but by porting it over to a new pkg) exists: https://github.com/cockroachdb/cockroach/blob/6ebb1726fadffbde44c4df4be0ffed965c06b1ff/pkg/ccl/logictestccl/testdata/logic_test/multi_region#L283-L293

if this was what you were thinking about


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140617)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140617#-OLAoOApBjNsMO76bYDP:-OLEthh2BzWhhak0ETDr:bwid77o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d13d41f499fdb7cf8c87a9677c805486de8becf9/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L128)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

re:nit -- ah oki right

re: broadness of `idx != nil` check -- i aligned this with legacy's implementation https://github.com/cockroachdb/cockroach/blob/42b9060871865023e97bbbe2e42ada6a8cec55d3/pkg/sql/add_column.go#L79-L92; iiuc this should still be equivalent since primary keys are also unique, but lmk if that's not what it going on here https://github.com/cockroachdb/cockroach/blob/4ab689e1d9f5420c6cef41de6808455ed55e2a14/pkg/sql/catalog/tabledesc/table.go#L42-L44
</blockquote></details>

that makes sense to me! thanks
___
*[`pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_regional_by_row/add_column_regional_by_row.definition` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140617#-OLApCX1BwCHULp_ZHIV:-OLEtlYCBj1NPsimqhdn:b-463b7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d13d41f499fdb7cf8c87a9677c805486de8becf9/pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_regional_by_row/add_column_regional_by_row.definition#L2)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

i feel like this is sufficient -- at least for a DSC end_to_end test, but what were you thinking about testing? the zone config generated from a `SURVIVE ZONE FAILURE` (which is code i touched, but by porting it over to a new pkg) exists: https://github.com/cockroachdb/cockroach/blob/6ebb1726fadffbde44c4df4be0ffed965c06b1ff/pkg/ccl/logictestccl/testdata/logic_test/multi_region#L283-L293

if this was what you were thinking about
</blockquote></details>

that is basically all i had in mind. i'm guessing the schema change plan isn't that different whether the database has REGION FAILURE or ZONE FAILURE then?


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140617)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_regional_by_row/add_column_regional_by_row.definition` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140617#-OLApCX1BwCHULp_ZHIV:-OLEuBW6ByOa5n5yl5A5:b-mjcxdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d13d41f499fdb7cf8c87a9677c805486de8becf9/pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_regional_by_row/add_column_regional_by_row.definition#L2)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

that is basically all i had in mind. i'm guessing the schema change plan isn't that different whether the database has REGION FAILURE or ZONE FAILURE then?
</blockquote></details>

nod nod (i verified this by rewriting the tests after modifying REGION FAILURE to ZONE FAILURE and looking at the diff as well -- the plan was the same)


<!-- Sent from Reviewable.io -->

annrpom commented:
test failure is unrelated to these changes:
```
    explain_test.go:207: pq: internal error: argument expression has type RECORD, need type USER DEFINED RECORD: public.seed, assignment cast isn't possible: WITH        	with_27 (col_270)
```

TFTR! ('-')7

bors r+
